<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134228602-6"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134228602-6');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>🧛🏻 👍🏾 🆎 Kotlin / Java错误处理：如何正确执行？ 🧑🏻‍🤝‍🧑🏻 🕜 💐</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="来源 


 错误处理在任何开发中都起着至关重要的作用。 程序中几乎所有内容都会出错：用户将输入错误的数据，或者它们可能通过http输入，或者在编写序列化/反序列化时出错，并且在处理过程中程序因错误而崩溃。 是的，它可能会浪费磁盘空间。 
 扰流板 

 ¯_（ツ）_ /¯，没有单一的方法，在每种特...">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <link href="https://fonts.googleapis.com/css?family=Quicksand&display=swap" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script>document.write('<script src="//pagea' + 'd2.googles' + 'yndication.com/pagea' + 'd/js/a' + 'dsby' + 'google.js"><\/script>')</script>
  <script>
        var superSpecialObject = {};
        superSpecialObject['google_a' + 'd_client'] = 'ca-p' + 'ub-6974184241884155';
        superSpecialObject['enable_page_level_a' + 'ds'] = true;
       (window['a' + 'dsbygoogle'] = window['a' + 'dsbygoogle'] || []).push(superSpecialObject);
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly-es.github.io/index.html"></a>
    <div class="page-header-text">Geekly articles weekly</div>
  </header>
  <section class="page js-page"><h1>Kotlin / Java错误处理：如何正确执行？</h1><div class="post__body post__body_full"><div class="post__text post__text-html" id="post-content-body" data-io-article-url="https://habr.com/ru/company/funcorp/blog/471766/"><img src="https://habrastorage.org/webt/hw/w3/mc/hww3mcjps5hwhykuwyzwj8upunu.jpeg"><br><p>  <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=zh-CN&amp;u=">来源</a> </p><br><p> 错误处理在任何开发中都起着至关重要的作用。 程序中几乎所有内容都会出错：用户将输入错误的数据，或者它们可能通过http输入，或者在编写序列化/反序列化时出错，并且在处理过程中程序因错误而崩溃。 是的，它可能会浪费磁盘空间。 </p><br><div class="spoiler">  <b class="spoiler_title">扰流板</b> <div class="spoiler_text"><p>  ¯_（ツ）_ /¯，没有单一的方法，在每种特定情况下，您都必须选择最合适的选项，但是对于如何做得更好，有一些建议。 </p></div></div><a name="habracut"></a><br><h3 id="predislovie"> 前言 </h3><br><p> 不幸的是（还是这样的生活？），这个清单还在继续。 开发人员经常需要考虑以下事实：错误可能在某处发生，并且有两种情况： </p><br><ul><li> 当在调用我们提供的函数并尝试进行处理时发生预期的错误时； </li><li> 当我们无法预见的操作过程中发生意外错误时。 </li></ul><br><p> 而且，如果预期的错误至少已本地化，那么其余错误几乎可以在任何地方发生。 如果我们不处理任何重要的事情，那么我们可能会因错误而崩溃（尽管此行为还不够，并且您至少需要在错误日志中添加一条消息）。 但是，如果现在付款正在处理中，您就不能拒绝，但是至少您需要返回有关操作失败的回复？ </p><br><p> 在介绍处理错误的方法之前，请先介绍一些有关Exception（异常）的信息： </p><br><h3 id="exception"> 例外情况 </h3><br><img src="https://habrastorage.org/webt/hn/mx/yz/hnmxyzwyfbelfe32oqi1nyite4o.png"><br><p>  <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=zh-CN&amp;u=">来源</a> </p><br><p>异常的层次结构已得到很好的描述，您可以找到很多有关异常的信息，因此在此处绘制异常是没有意义的。 有时仍引起激烈讨论的是已<code>checked</code>和<code>unchecked</code>错误。 而且尽管大多数人都接受<code>unchecked</code>例外作为首选（在Kotlin中根本没有<code>checked</code>例外），但并非所有人都对此表示赞同。 </p><br><p> 被<code>checked</code>异常确实有一个很好的意图，使它们成为一种方便的错误处理机制，但是现实进行了调整，尽管将可以从此函数抛出的所有异常引入签名中的想法是可以理解和逻辑的。 </p><br><p> 让我们来看一个例子。 假设我们有一个<code>method</code>函数可以抛出一个经过检查的<code>PanicException</code> 。 这样的功能如下所示： </p><br><pre> <code class="java hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">public</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">void</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">method</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span><span class="hljs-function"> </span><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">throws</span></span></span><span class="hljs-function"> PanicException </span></span>{ }</code> </pre> <br><p> 从她的描述中很明显，她可以抛出一个异常，并且只能有一个异常。 看起来很舒服吗？ 尽管我们有一个小程序，仅此而已。 但是，如果程序稍大，并且有更多这样的功能，则会出现一些问题。 </p><br><p> 根据规范，检查异常要求在功能签名中列出<strong>所有可能的</strong>检查异常（或它们的共同祖先）。 因此，如果我们有一连串的调用a-&gt; <code>b</code> &gt; <code>c</code>并且嵌套最多的函数抛出某种异常，那么应该为该链中的每个人都放下该异常。 并且，如果有几个例外，则签名中最顶层的函数应具有所有例外的描述。 </p><br><p> 因此，随着程序变得更加复杂，这种方法导致这样一个事实，即顶层函数的异常逐渐向普通祖先崩溃，并最终下降为<code>Exception</code> 。 这种形式的内容类似于<code>unchecked</code>异常，并且抵消了检查的异常的所有优点。 </p><br><p> 鉴于该程序作为一种活生物体正在不断变化和发展，因此几乎不可能事先预见该程序中可能会出现什么例外情况。 结果是，当我们添加一个带有新异常的新功能时，我们必须遍历整个使用过程，并更改所有功能的签名。 同意，这不是最令人愉快的任务（即使考虑到现代IDE为我们做到这一点）。 </p><br><p> 但是最后一个，也可能是最大的检查异常是从Java 8“驱使” lambda。签名中没有检查异常__（ツ）_ /（因为可以在lambda中调用任何函数，而任何签名），因此任何来自lambda的具有已检查异常的函数调用都将其包装为未检查的异常转发中： </p><br><pre> <code class="java hljs">Stream.of(<span class="hljs-number"><span class="hljs-number">1</span></span>,<span class="hljs-number"><span class="hljs-number">2</span></span>,<span class="hljs-number"><span class="hljs-number">3</span></span>).forEach(item -&gt; { <span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { functionWithCheckedException(); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (Exception e) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> RuntimeException(<span class="hljs-string"><span class="hljs-string">"rethrow"</span></span>, e); } });</code> </pre> <br><p> 幸运的是，在JVM规范中根本没有经过检查的异常，因此在Kotlin中，您不能在同一个lambda中包装任何内容，而只是调用所需的函数。 </p><br><div class="spoiler">  <b class="spoiler_title">虽然有时...</b> <div class="spoiler_text"><p> 尽管有时这会导致意想不到的后果，例如， <code>Spring Framework</code>中<code>@Transactional</code>的错误操作，该错误仅“期待”异常情况。 但这更多是框架的功能，也许在不久的将来的<a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=zh-CN&amp;u=">github问题中</a> ，Spring的这种行为将会改变。 </p></div></div><br><p> 异常本身是特殊对象。 除了可以通过方法“抛出”这些事实外，它们还在创建时收集stacktrace。 然后，此功能有助于分析问题和查找错误，但是如果应用程序逻辑与抛出的异常紧密联系在一起，则还可能导致某些性能问题。 如该<a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=zh-CN&amp;u=">文章</a>所示，在这种情况下，禁用stacktrace程序集可以显着提高其性能，但是只有在确实需要它的特殊情况下，才应诉诸于此！ </p><br><h3 id="obrabotka-oshibok"> 错误处理 </h3><br><p> 处理“意外”错误的主要方法是找到一个可以拦截它们的地方。 在JVM语言中，这可以是http方法的流创建点或过滤器/入口点，您可以在其中放置try-catch来处理<code>unchecked</code>错误。 如果使用任何框架，则很可能已经具有创建常见错误处理程序的能力，例如在Spring框架中，可以使用带有<code>@ExceptionHandler</code>批注的方法。 </p><br><p> 您可以通过抛出相同的<code>unckecked</code>处理的异常来“引发”我们不想在特定位置处理的这些中央处理点的异常（例如，当我们不知道在特定位置做什么以及如何处理错误时）。 但是这种方法并不总是合适的，因为有时可能需要处理错误，并且您需要检查是否正确处理了函数调用的所有位置。 考虑执行此操作的方法。 </p><br><ol><li><p> 仍然使用异常和相同的try-catch： </p><br><pre> <code class="java hljs"> <span class="hljs-keyword"><span class="hljs-keyword">int</span></span> a = <span class="hljs-number"><span class="hljs-number">10</span></span>; <span class="hljs-keyword"><span class="hljs-keyword">int</span></span> b = <span class="hljs-number"><span class="hljs-number">20</span></span>; <span class="hljs-keyword"><span class="hljs-keyword">int</span></span> sum; <span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { sum = calculateSum(a,b); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (Exception e) { sum = -<span class="hljs-number"><span class="hljs-number">1</span></span>; }</code> </pre> <br><p> 主要缺点是，我们可以“忘记”在调用位置将其包装在try-catch中，而跳过尝试就地处理它的尝试，因此，异常将被抛出到错误处理的常见点。 在这里，我们可以转到<code>checked</code>异常（对于Java），但是然后我们将获得上面提到的所有缺点。 如果并非总是需要就地进行错误处理，但是在极少数情况下需要使用此方法，则使用此方法很方便。 </p><br></li><li><p> 使用密封类作为调用的结果（Kotlin）。 <br> 在Kotlin中，您可以限制类继承人的数量，使其在编译阶段可计算-这使编译器可以验证代码中是否解析了所有可能的选项。 在Java中，您可以创建一个公共接口和多个后代，但是会丢失编译级检查。 </p><br><pre> <code class="kotlin hljs"><span class="hljs-keyword"><span class="hljs-keyword">sealed</span></span> <span class="hljs-class"><span class="hljs-keyword"><span class="hljs-class"><span class="hljs-keyword">class</span></span></span><span class="hljs-class"> </span><span class="hljs-title"><span class="hljs-class"><span class="hljs-title">Result</span></span></span><span class="hljs-class"> </span><span class="hljs-title"><span class="hljs-class"><span class="hljs-title">data</span></span></span><span class="hljs-class"> </span><span class="hljs-title"><span class="hljs-class"><span class="hljs-title">class</span></span></span><span class="hljs-class"> </span><span class="hljs-title"><span class="hljs-class"><span class="hljs-title">SuccessResult</span></span></span></span>(<span class="hljs-keyword"><span class="hljs-keyword">val</span></span> value: <span class="hljs-built_in"><span class="hljs-built_in">Int</span></span>): Result() <span class="hljs-keyword"><span class="hljs-keyword">data</span></span> <span class="hljs-class"><span class="hljs-keyword"><span class="hljs-class"><span class="hljs-keyword">class</span></span></span><span class="hljs-class"> </span><span class="hljs-title"><span class="hljs-class"><span class="hljs-title">ExceptionResult</span></span></span></span>(<span class="hljs-keyword"><span class="hljs-keyword">val</span></span> exception: Exception): Result() <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> a = <span class="hljs-number"><span class="hljs-number">10</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> b = <span class="hljs-number"><span class="hljs-number">20</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> sum = <span class="hljs-keyword"><span class="hljs-keyword">when</span></span> (<span class="hljs-keyword"><span class="hljs-keyword">val</span></span> result = calculateSum(a,b)) { <span class="hljs-keyword"><span class="hljs-keyword">is</span></span> SuccessResult -&gt; result.value <span class="hljs-keyword"><span class="hljs-keyword">is</span></span> ExceptionResult -&gt; { result.exception.printStackTrace() -<span class="hljs-number"><span class="hljs-number">1</span></span> } }</code> </pre> <br><p> 当您需要显式检查结果值（或显式忽略）时，这里得到类似于<code>golang</code>错误方法的信息。 该方法非常实用，在每种情况下都需要抛出很多参数时特别方便。 可以使用各种方法扩展<code>Result</code>类，使其更容易获得上面抛出异常的结果（如果有的话）（即，我们不需要在调用位置处理错误）。 主要缺点将仅是创建中间多余的对象（以及稍微冗长的条目），但也可以使用<code>inline</code>类将其删除（如果一个参数足以满足我们的需要）。 作为一个特定示例，还有Kotlin的<code>Result</code>类。 没错，它仅供内部使用，因为 将来，其实现可能会稍有变化，但是如果要使用它，可以添加编译标志<code>-Xallow-result-return-type</code> 。 </p><br></li><li><p> 作为权利要求2的一种可能的类型，使用<code>Either</code>的功能编程中的类型，它可以是结果，也可以是错误。 类型本身可以是<code>sealed</code>类或<code>inline</code>类。 以下是使用<code>arrow</code>库中的实现的示例： </p><br><pre> <code class="kotlin hljs"><span class="hljs-keyword"><span class="hljs-keyword">val</span></span> a = <span class="hljs-number"><span class="hljs-number">10</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> b = <span class="hljs-number"><span class="hljs-number">20</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> value = <span class="hljs-keyword"><span class="hljs-keyword">when</span></span>(<span class="hljs-keyword"><span class="hljs-keyword">val</span></span> result = calculateSum(a,b)) { <span class="hljs-keyword"><span class="hljs-keyword">is</span></span> Either.Left -&gt; { result.a.printStackTrace() -<span class="hljs-number"><span class="hljs-number">1</span></span> } <span class="hljs-keyword"><span class="hljs-keyword">is</span></span> Either.Right -&gt; result.b }</code> </pre> <br><p>  <code>Either</code>最适合喜欢功能性方法并且喜欢建立呼叫链的人员。 </p><br></li><li><p> 使用Kotlin中的<code>Option</code>或<code>nullable</code>类型： </p><br><pre> <code class="kotlin hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">fun</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">testFun</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span></span> { <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> a = <span class="hljs-number"><span class="hljs-number">10</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> b = <span class="hljs-number"><span class="hljs-number">20</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> sum = calculateSum(a,b) ?: <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> RuntimeException(<span class="hljs-string"><span class="hljs-string">"some exception"</span></span>) } <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">fun</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">calculateSum</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">(a: </span></span><span class="hljs-type"><span class="hljs-function"><span class="hljs-params"><span class="hljs-type">Int</span></span></span></span><span class="hljs-function"><span class="hljs-params">, b: </span></span><span class="hljs-type"><span class="hljs-function"><span class="hljs-params"><span class="hljs-type">Int</span></span></span></span><span class="hljs-function"><span class="hljs-params">)</span></span></span></span>: <span class="hljs-built_in"><span class="hljs-built_in">Int</span></span>?</code> </pre> <br><p> 如果错误原因不是很重要，并且仅是一个原因，则此方法适用。 空答案被认为是错误，并被抛出更高的值。 最短的记录，没有创建其他对象，但是这种方法不能总是被应用。 </p><br></li><li><p> 与第4项类似，仅将硬编码值用作错误标记： </p><br><pre> <code class="kotlin hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">fun</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">testFun</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">()</span></span></span></span> { <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> a = <span class="hljs-number"><span class="hljs-number">10</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> b = <span class="hljs-number"><span class="hljs-number">20</span></span> <span class="hljs-keyword"><span class="hljs-keyword">val</span></span> sum = calculateSum(a,b) <span class="hljs-keyword"><span class="hljs-keyword">if</span></span> (sum == -<span class="hljs-number"><span class="hljs-number">1</span></span>) { <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> RuntimeException(“error”) } } <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">fun</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">calculateSum</span></span></span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">(a: </span></span><span class="hljs-type"><span class="hljs-function"><span class="hljs-params"><span class="hljs-type">Int</span></span></span></span><span class="hljs-function"><span class="hljs-params">, b: </span></span><span class="hljs-type"><span class="hljs-function"><span class="hljs-params"><span class="hljs-type">Int</span></span></span></span><span class="hljs-function"><span class="hljs-params">)</span></span></span></span>: <span class="hljs-built_in"><span class="hljs-built_in">Int</span></span></code> </pre> <br><p> 这可能是<code>C</code> （甚至Algol）提供的最古老的错误处理方法。 没有开销，只有一个不完全清楚的代码（以及对结果选择的限制），但是与第4段不同，如果需要多个可能的异常，则可以制作各种错误代码。 </p><br></li></ol><br><h3 id="vyvody"> 结论 </h3><br><p> 可以根据情况将所有方法组合在一起，其中没有一种方法适合所有情况。 </p><br><p> 因此，例如，您可以使用<code>sealed</code>类实现<code>golang</code>处理错误的方法，如果不是很方便，则转到<code>unchecked</code>错误。 </p><br><p> 或者，在大多数地方， <code>nullable</code>为<code>nullable</code>类型作为无法计算值或无法从某处获取值的标记（例如，作为表明在数据库中找不到值的指示符）。 </p><br><p> 并且，如果您具有<code>arrow</code>或其他类似库的功能齐全的代码，则最有可能最好使用<code>Either</code> 。 </p><br><p> 对于http服务器，最容易将所有错误提高到中心点，并且仅在某些地方将<code>nullable</code>方法与<code>sealed</code>类结合使用。 </p><br><p> 我会很高兴在评论中看到您正在使用此功能，或者还有其他便捷的错误处理方法？ </p><br><p> 并感谢阅读到底的每个人！ </p></div></div><p>Source: <a rel="nofollow" href="https://habr.com/ru/post/zh-CN471766/">https://habr.com/ru/post/zh-CN471766/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../zh-CN471748/index.html">药房优化：我们对数学的处理</a></li>
<li><a href="../zh-CN471750/index.html">特权访问管理是信息安全中的优先任务（例如，Fudo PAM）</a></li>
<li><a href="../zh-CN471756/index.html">服务器如何启动</a></li>
<li><a href="../zh-CN471758/index.html">如何将文字翻译成其他语言？</a></li>
<li><a href="../zh-CN471760/index.html">爱德华·斯诺登：战场-加密</a></li>
<li><a href="../zh-CN471770/index.html">认识大数据工具：IntelliJ IDEA中的Spark和Zeppelin笔记本支持</a></li>
<li><a href="../zh-CN471772/index.html">用于启动和开发Java应用程序，在JVM上进行编译，执行的工具</a></li>
<li><a href="../zh-CN471774/index.html">茶壶中的Android Camera2 API，第2部分，编写视频</a></li>
<li><a href="../zh-CN471776/index.html">Laravel：解释基本概念。 第二部分：练习</a></li>
<li><a href="../zh-CN471778/index.html">在HackQuest 2019上打破Micosoft Lunix</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter57283870 = new Ya.Metrika({
                  id:57283870,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/57283870" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134228602-6', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

<footer class="page-footer">
  <div class="page-footer-legal-info-container page-footer-element">
    <p>
      Weekly-Geekly ES | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
    </p>
  </div>
  <div class="page-footer-counters-container page-footer-element">
    <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=9uU9J9pq8z7k8xEBHYSfs6DenIBAHs3vLIHcPIJW9d0&co=3a3a3a&ct=ffffff'/></a>
  </div>
</footer>
  
</body>

</html>